tools/fast/series/wait_for_chain_message.go

```diff
@@ -0,0 +1,67 @@
+package series
+
+import (
+	"context"
+	"io"
+
+	"github.com/filecoin-project/go-filecoin/tools/fast"
+	"github.com/filecoin-project/go-filecoin/types"
+	"github.com/ipfs/go-cid"
+)
+
+// MsgInfo contains the BlockCid for the message MsgCid
+type MsgInfo struct {
+	BlockCid cid.Cid
+	MsgCid   cid.Cid
+}
+
+// WaitForChainMessage iterates over the chain until the provided function `fn` returns true.
+func WaitForChainMessage(ctx context.Context, node *fast.Filecoin, fn func(context.Context, *fast.Filecoin, *types.SignedMessage) (bool, error)) (*MsgInfo, error) {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-CtxSleepDelay(ctx):
+			dec, err := node.ChainLs(ctx)
+			if err != nil {
+				return nil, err
+			}
+
+			for dec.More() {
+				var blks []types.Block
+				err := dec.Decode(&blks)
+				if err != nil {
+					if err == io.EOF {
+						break
+					}
+
+					return nil, err
+				}
+
+				for _, blk := range blks {
+					msgs, err := node.ShowMessages(ctx, blk.Messages)
+					if err != nil {
+						return nil, err
+					}
+
+					for _, msg := range msgs {
```

Comment:
A quadruple for-loop, while fun to say, isn't fun to read. Can we extract a few of these :loop::loop: into methods?